### PR TITLE
fix: fix checkpoint gc which was incorrectly deleting checkpoints [DET-9702]

### PR DESCRIPTION
**Pull request by salonig23**
## Description

Checkpoint garbage collection was incorrectly deleting all the checkpoints except the checkpoint from the latest trial. It was fixed by changing the sql query that determines the checkpoints to be deleted.

## Test Plan

Create a few experiments with different configurations for garbage collection, and make sure the checkpoints saved are consistent with those configurations.

## Checklist

- [ ] Changes have been manually QA'd
- [ ] User-facing API changes need the "User-facing API Change" label.
- [ ] Release notes should be added as a separate file under `docs/release-notes/`.
  See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.
- [ ] Licenses should be included for new code which was copied and/or modified from any external code.

## Ticket

DET-9702

### CODE REVIEW

File: master/internal/db/postgres_experiments.go [hunk 1]
```diff
@@ -1108,74 +1108,57 @@ func (db *PgDB) ExperimentCheckpointsToGCRaw(
...
-    ) c, const
-    WHERE (SELECT COUNT(*) FROM trials t WHERE t.warm_start_checkpoint_id = c.id) = 0
-          AND c.trial_order_rank > $4
-          AND ((c.experiment_rank > $2
-                AND c.trial_rank > $3)
-               OR (c.step->'validation'->'metrics'->'validation_metrics'->>const.metric_name
-                   IS NULL))
-)
-SELECT selected_checkpoints.uuid AS ID from selected_checkpoints;`
+	WITH const AS (
```
Comment (stoksc) [inline master/internal/db/postgres_experiments.go:1111]: can you unindent all these so the diff is more reviewable?

Comment (salonig23) [reply master/internal/db/postgres_experiments.go]: Done! I restructured the query so some parts of the query that are the same will still show as diff unfortunately

File: master/internal/db/postgres_experiments.go [hunk 2]
```diff
@@ -1117,65 +1117,48 @@ WITH const AS (
...
+		FROM (
+			SELECT c.id as c_id, c.trial_id, c.state,
+				c.steps_completed as total_batches, c.report_time as end_time, c.uuid, 
+				c.resources, c.metadata
+			FROM checkpoints_view c, trials t
+			WHERE c.state = 'COMPLETED' AND c.trial_id = t.id AND t.experiment_id = $1 
+		) as ct
+		LEFT JOIN validations v ON v.trial_id = ct.trial_id AND v.total_batches = ct.total_batches
+	) AS sub, const
+	WHERE (SELECT COUNT(*) FROM trials t WHERE t.warm_start_checkpoint_id = sub.c_id) = 0)
```
Comment (stoksc) [inline master/internal/db/postgres_experiments.go:1158]: it doesn't make sense to me that making a checkpoint a warm start checkpoint would affect how all the others are GC'd due to `rank()` changes. can we move this to the `WHERE` attached to `FROM selected_checkpoints`.

**Review by stoksc — COMMENTED** [review]
looks great, really. just need a test, and i have a lot of feedback to try to make this query not one of the most arduous in the system

File: master/internal/db/postgres_experiments_intg_test.go [hunk 3]
```diff
@@ -54,6 +55,12 @@ func TestExperimentCheckpointsToGCRaw(t *testing.T) {
 		0,
 	)
 	require.NoError(t, err)
+	sort.Slice(expectedCheckpoints, func(i, j int) bool {
+		return expectedCheckpoints[i].String() < expectedCheckpoints[j].String()
```
[the remark this answers is not in the document]
Comment (salonig23) [reply master/internal/db/postgres_experiments_intg_test.go:59]: Done!

File: master/internal/db/postgres_experiments.go [hunk 4]
```diff
@@ -1117,65 +1117,38 @@ WITH const AS (
...
+			PARTITION BY trial_id
+			ORDER BY const.sign * (v.metrics->'validation_metrics'->>const.metric_name)::float8 
+			ASC NULLS LAST, v.id ASC
+		) AS trial_rank,
+		rank() OVER (
+			PARTITION BY trial_id
+			ORDER BY (c.metadata->>'steps_completed')::int DESC
+		) AS trial_order_rank
+	FROM checkpoints_v2 c
+	NATURAL JOIN const
```
Comment (stoksc) [inline master/internal/db/postgres_experiments.go:1140]: i know i helped write this but it is probably clearer as
```suggestion
	JOIN const ON true
```

Comment (stoksc) [reply master/internal/db/postgres_experiments.go:1140]: i played around with doing it with `CROSS JOIN` and `LEFT JOIN LATERAL` for fun. all possible, all emit similar plans and this is the most clear to me.

File: master/internal/db/postgres_experiments_intg_test.go [hunk 5]
```diff
@@ -54,7 +62,29 @@ func TestExperimentCheckpointsToGCRaw(t *testing.T) {
 		0,
 	)
 	require.NoError(t, err)
-	require.Equal(t, expectedCheckpoints, checkpoints)
+	sort.Slice(expectedCheckpoints1, func(i, j int) bool {
```
Comment (stoksc) [inline master/internal/db/postgres_experiments_intg_test.go:65]: could you split this into two tests instead or, even better, use table driven tests (https://dave.cheney.net/2019/05/07/prefer-table-driven-tests)? as a rule of thumb, one test should test one thing; it is much easier to maintain the tests then (especially understand them when failures happen).

File: master/internal/db/postgres_test_utils.go [hunk 6]
```diff
@@ -319,6 +322,54 @@ func MockModelCheckpoint(
 	return ckpt
 }
 
+// MockModelCheckpointSteps returns a mock model checkpoint with specified steps.
+func MockModelCheckpointSteps(
+	ckptUUID uuid.UUID, tr *model.Trial, a *model.Allocation, stepsCompleted int,
+) model.CheckpointV2 {
```
Comment (stoksc) [inline master/internal/db/postgres_test_utils.go:328]: could add functional opts to `MockModelCheckpoint` instead and save writing an entire new function

**Review by stoksc — COMMENTED** [review]
only comments are stylist stuff with tests. lmk if you would like to pair for 30m on it tomorrow and then we can get this in.

File: master/internal/db/postgres_experiments.go (hunk 4 above)
Comment (stoksc) [reply master/internal/db/postgres_experiments.go:1140]: can we go ahead and pull this in?

Comment (stoksc) [reply master/internal/db/postgres_experiments.go:1140]: oh we merged.

**Pull request merged**